File: clients/ios/Lextures/Core/Routing/MobileIaPreferences.swift
import Foundation

/// Client-side rollout flag for the M0.5 information-architecture redesign.
/// Default off so the legacy 5-tab shell remains the rollback path (AC-6).
enum MobileIaPreferences {
    private static let key = "mobile_ia_redesign"
    private static let universalSearchKey = "mobile_universal_search"

    static var isRedesignEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: key) }
        set { UserDefaults.standard.set(newValue, forKey: key) }
    }

    static var isUniversalSearchEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: universalSearchKey) }
        set { UserDefaults.standard.set(newValue, forKey: universalSearchKey) }
    }

    private static let roleContextKey = "mobile_ia_role_context"

    static func loadRoleContext() -> MobileRoleContext? {
        guard let raw = UserDefaults.standard.string(forKey: roleContextKey) else { return nil }
        return MobileRoleContext(rawValue: raw)
    }

    static func saveRoleContext(_ context: MobileRoleContext) {
        UserDefaults.standard.set(context.rawValue, forKey: roleContextKey)
    }

    private static let selectedChildKey = "mobile_parent_selected_child"

    static func loadSelectedChildId() -> String? {
        UserDefaults.standard.string(forKey: selectedChildKey)
    }

    static func saveSelectedChildId(_ studentId: String) {
        UserDefaults.standard.set(studentId, forKey: selectedChildKey)
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/navigation/MobileIaPreferences.kt
package com.lextures.android.core.navigation

import android.content.Context

/** Client-side rollout flag for the M0.5 information-architecture redesign. */
object MobileIaPreferences {
    private const val PREFS = "mobile_ia_prefs"
    private const val KEY_REDESIGN = "mobile_ia_redesign"
    private const val KEY_UNIVERSAL_SEARCH = "mobile_universal_search"
    private const val KEY_ROLE_CONTEXT = "mobile_ia_role_context"

    fun isRedesignEnabled(context: Context): Boolean =
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getBoolean(KEY_REDESIGN, false)

    fun setRedesignEnabled(context: Context, enabled: Boolean) {
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_REDESIGN, enabled)
            .apply()
    }

    fun isUniversalSearchEnabled(context: Context): Boolean =
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getBoolean(KEY_UNIVERSAL_SEARCH, false)

    fun setUniversalSearchEnabled(context: Context, enabled: Boolean) {
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putBoolean(KEY_UNIVERSAL_SEARCH, enabled)
            .apply()
    }

    fun loadRoleContext(context: Context): MobileRoleContext? {
        val raw = context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_ROLE_CONTEXT, null)
            ?: return null
        return runCatching { MobileRoleContext.valueOf(raw) }.getOrNull()
    }

    fun saveRoleContext(context: Context, roleContext: MobileRoleContext) {
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY_ROLE_CONTEXT, roleContext.name)
            .apply()
    }

    private const val KEY_SELECTED_CHILD = "mobile_parent_selected_child"

    fun loadSelectedChildId(context: Context): String? =
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .getString(KEY_SELECTED_CHILD, null)

    fun saveSelectedChildId(context: Context, studentId: String) {
        context.applicationContext
            .getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit()
            .putString(KEY_SELECTED_CHILD, studentId)
            .apply()
    }
}